File: backend/app/schemas/dataset.py
from datetime import datetime

from pydantic import BaseModel, Field


class DatasetColumnResponse(BaseModel):
    name: str
    position: int
    inferred_type: str
    null_count: int
    sample_values: list | None = None


class DatasetResponse(BaseModel):
    id: str
    name: str
    original_filename: str
    file_type: str
    row_count: int
    column_count: int
    status: str
    file_hash: str | None = None
    error_message: str | None = None
    columns: list[DatasetColumnResponse] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DatasetListItem(BaseModel):
    id: str
    name: str
    original_filename: str
    file_type: str
    row_count: int
    column_count: int
    status: str
    created_at: datetime


class DatasetPreviewRow(BaseModel):
    row_index: int
    data: dict


class DatasetPreviewResponse(BaseModel):
    dataset_id: str
    rows: list[DatasetPreviewRow]
    total_rows: int


class DatasetListResponse(BaseModel):
    items: list[DatasetListItem]
    total: int
    page: int
    page_size: int


class ApiResponse(BaseModel):
    success: bool = True
    data: dict | list | None = None
    error: dict | None = None

File: frontend/src/types/dataset.ts
export interface DatasetColumn {
  name: string;
  position: number;
  inferred_type: string;
  null_count: number;
  sample_values: unknown[] | null;
}

export interface Dataset {
  id: string;
  name: string;
  original_filename: string;
  file_type: string;
  row_count: number;
  column_count: number;
  status: string;
  file_hash: string | null;
  error_message: string | null;
  columns: DatasetColumn[];
  created_at: string;
  updated_at: string;
}

export interface DatasetListItem {
  id: string;
  name: string;
  original_filename: string;
  file_type: string;
  row_count: number;
  column_count: number;
  status: string;
  created_at: string;
}

export interface DatasetListResponse {
  items: DatasetListItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface DatasetPreviewResponse {
  dataset_id: string;
  rows: { row_index: number; data: Record<string, unknown> }[];
  total_rows: number;
}

export interface ApiResponse<T = unknown> {
  success: boolean;
  data: T | null;
  error: { code: string; message: string } | null;
}
